### PR TITLE
Add initial Sphinx docs

### PR DESCRIPTION
**Pull request by webknjaz**
PR preview: https://proxypy--747.org.readthedocs.build/en/747/

## Notes

* Extended Markdown support in Sphinx is implemented via MyST. Start learning about it here: https://myst-parser.readthedocs.io/en/latest/sphinx/intro.html#extend-markdown-with-a-directive.
* The docstrings must use RST, there's no way around it. There's a linter called darglint that is integrated with flake8. It is currently set up to require the Sphinx's built-in docstring markup. Although, its rules are disabled atm because of countless violations. If you want to switch to a different style for documenting function/class signatures, you need to change it @ https://github.com/abhinavsingh/proxy.py/blob/develop/.darglint
* Most of the modules have a copyright notice in their module docstrings. I recommend changing that to a comment block because those docstring get into the private API per-module pages and there's a lot of pointless duplication because of this.

### CODE REVIEW

File: docs/spelling_wordlist.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+...
```
Comment (webknjaz) [inline docs/spelling_wordlist.txt:1]: This is a dictionary for the words libenchant doesn't know about.

**Comment by abhinavsingh** [comment]
Can you try this

```
...
Changelog
cloudflare
Cloudflare
DoH
IPv
Nginx
Pluggable
Pre
Realtime
Scalable
Submodules
Subpackages
Teardown
TestCase
Threadless
Url
WebSocket
Websocket
acceptor
acceptors
api
auth
compat
conn
csr
datastructure
decrypt
decrypted
dev
devtools
discoverable
dns
doesn
eventing
facebook
filedescriptor
fileno
hardcoded
html
http
httpbin
httpx
imoracle
instagram
ip
iterable
klass
linkedin
localhost
macOS
malwares
memoryview
nd
pac
pki
pre
py
reusability
scm
shortlink
sys
tcp
teardown
threadless
unittest
url
urllib
urlparse
utils
websocket
whatsapp
ws
youtube
```

Following worked for most scenarios:

```console
$ tox -e spellcheck-docs > spell.txt 2>&1 
$ cat spell.txt | grep "^/" | grep "Spell check:" | awk '{ print substr($4, 1, length($4)-1) }' | sort | uniq`
```

`spellcheck-docs` passed:

```
writing output... [100%] pkg/proxy.proxy                                                                          
build succeeded.
___________________________________ summary ____________________________________
  spellcheck-docs: commands succeeded
  congratulations :)
```

File: docs/spelling_wordlist.txt (hunk 1 above)
Comment (abhinavsingh) [reply docs/spelling_wordlist.txt:1]: Shared a list of words in the comments.  Thank you for this PR and addressing several existing doc scenarios.  I also see some `requirements.in` and `requirements.txt` magic going on.  I haven't used it before, something to dig into later.

**Comment by webknjaz** [comment]
Sure, I just didn't have time to do it.

**Comment by webknjaz** [comment]
@abhinavsingh looks like you didn't fully integrate RTD into this repo — I don't see checks reported in this PR.

File: docs/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,272 @@
+#
+# This file is autogenerated by pip-compile with python 3.10
+# To update, run:
+#
+#    pip-compile --allow-unsafe --generate-hashes --output-file=docs/requirements.txt --strip-extras docs/requirements.in
```
Comment (webknjaz) [inline docs/requirements.txt:5]: @abhinavsingh note that this file is auto-generated and is checked in for reproducibility purposes.

Comment (abhinavsingh) [reply docs/requirements.txt:5]: Thank you, I'll spend sometime to try this out.  I haven't used this option before, but I see where it's coming from.  I have now about a dozen TODOs on my list, thanks to you :D

**Comment by abhinavsingh** [comment]
> Sure, I just didn't have time to do it.

No worries, happy to help.

**Comment by abhinavsingh** [comment]
> @abhinavsingh looks like you didn't fully integrate RTD into this repo — I don't see checks reported in this PR.

Missed this comment

```
/home/docs/checkouts/readthedocs.org/user_builds/proxypy/envs/latest/bin/python -m mkdocs build --clean --site-dir _build/html --config-file mkdocs.yml
ERROR   -  Config value: 'site_dir'. Error: The 'site_dir' should not be within the 'docs_dir' as this leads to the build directory being copied into itself and duplicate nested files in the 'site_dir'.(site_dir: '/home/docs/checkouts/readthedocs.org/user_builds/proxypy/checkouts/latest/_build/html', docs_dir: '/home/docs/checkouts/readthedocs.org/user_builds/proxypy/checkouts/latest') 

Aborted with 1 Configuration Errors!
```

I see following, but may be because it's trying `mkdocs`?  I changed it to `Sphinx Htmldir` as you suggested.

**Comment by abhinavsingh** [comment]
Also enabled `build pull request` which was disabled.

**Comment by webknjaz** [comment]
> I see following, but may be because it's trying `mkdocs`? I changed it to `Sphinx Htmldir` as you suggested.

Yes, mkdocs is not Sphinx.


> Also enabled `build pull request` which was disabled.

I think you could also add it to the branch protection settings.

**Comment by webknjaz** [comment]
> > I see following, but may be because it's trying `mkdocs`? I changed it to `Sphinx Htmldir` as you suggested.
> 
> Yes, mkdocs is not Sphinx.
> 
> > Also enabled `build pull request` which was disabled.
> 
> I think you could also add it to the branch protection settings.

Especially taking into account that build is successful on RTD (https://readthedocs.org/projects/proxypy/builds/15307154/) and the preview is here https://proxypy--747.org.readthedocs.build/en/747/.

**Comment by abhinavsingh** [comment]
This is awesome.  I am tempted to take a pass at end-to-end doc string once again :)

I am thinking of 2 edits to make the site more accessible:

1) Customize these top-level names.  I understand these are module name, but probably we are better off by just reproducing the directory structure e.g.

```
- proxy.py
  - common
  - core
  - ...
```

2) Can we somehow auto-expand the menu and keep it above `Glossary`.  I plan to add some tutorial sections, now that we have a doc site, but I think this still belongs up there.  Wdyt?   I had to look around for where are the class doc strings :)

**Comment by abhinavsingh** [comment]
> * Most of the modules have a copyright notice in their module docstrings. I recommend changing that to a comment block because those docstring get into the private API per-module pages and there's a lot of pointless duplication because of this.

+1. I noticed the same, redundant header on every page.  Will make a pass at it and push a fix.

**Comment by webknjaz** [comment]
@abhinavsingh you don't need to customize the auto-generated **private** API pages. It's best to add more human-readable at separate pages (starting with splitting the README into several pages).
But if you really want to, you could try changing the templates as documented here https://www.sphinx-doc.org/en/master/man/sphinx-apidoc.html#cmdoption-sphinx-apidoc-0.

On the second note, I recommend checking out the https://diataxis.fr framework — it is a set of principles for structuring the docs sections and other aspects of the content.

And I recommend joining https://www.writethedocs.org/slack/ — it's a community of mostly tech writers but there are channels for Sphinx and Diataxis where you can ask questions and get insights about improving the docs.

As for expanding or putting the Private API on top — don't. It's there mostly for convenience. Many sites hide it or don't create at all. You need to focus on adding more documents, once this is merged. The idea is that you'll add new document names to one of the ToC sections or create additional one at the top and that will show up above all other menu items.

File: README.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 ⚡ Fast • 🪶 Lightweight • 0️⃣ Dependency • 🔌 Pluggable • 😈 TLS interception • 🔒 DNS-over-HTTPS • 🔥 Poor Man's VPN • ⏪ Reverse & ⏩ Forward • 👮🏿 "Proxy Server" framework • 🌐 "Web Server" framework • ➵ ➶ ➷ ➠ "PubSub" framework • 👷 "Work" acceptor & executor framework
 </p>
 
+[//]: # (DO-NOT-REMOVE-docs-badges-START)
```
Comment (webknjaz) [inline README.md:7]: @abhinavsingh these markers are referenced from `docs/index.md`. They are used to include parts of the README on the front page, excluding the badges. It's enough for a demo now but you can make more and include things on other (sub)pages later.

File: README.md [hunk 4]
```diff
@@ -969,7 +973,7 @@ response from the server. Start `proxy.py` as:
     --ca-signing-key-file ca-signing-key.pem
 ```
 
-[![NOTE](https://img.shields.io/static/v1?label=MacOS&message=note&color=yellow)](https://github.com/abhinavsingh/proxy.py#flags) Also provide explicit CA bundle path needed for validation of peer certificates. See `--ca-file` flag.
+[![NOTE](https://img.shields.io/static/v1?label=MacOS&message=note&color=yellow)](https://github.com/abhinavsingh/proxy.py#user-content-flags) Also provide explicit CA bundle path needed for validation of peer certificates. See `--ca-file` flag.
```
Comment (webknjaz) [inline README.md:976]: This one is tricky. The `#flags` anchor doesn't exist in HTML, it's navigated to via JS so the linkcheck builder cannot verify that the link is valid. All of the real anchors are prefixed, this is why this patch was needed.

Comment (abhinavsingh) [reply README.md:976]: Thank you, I did notice this change and didn't understand the context.  Makes sense coz sphinx kind of steals away ToC from the readme.

File: proxy/core/acceptor/threadless.py [hunk 5]
```diff
@@ -37,19 +41,23 @@ class Threadless(multiprocessing.Process):
...
-    Threadless will then invoke necessary lifecycle of the `Work` protocol,
+    Threadless will then invoke necessary lifecycle of the
+    :class:`~proxy.core.acceptor.work.Work` protocol,
     allowing `work_klass` implementation to handle the assigned work.
 
-    Example, `BaseTcpServerHandler` implements `Work` protocol. It expects
-    a client connection as work payload and hooks into the threadless
-    event loop to handle the client connection.
+    Example, :class:`~proxy.core.base.tcp_server.BaseTcpServerHandler`
+    implements :class:`~proxy.core.acceptor.work.Work` protocol. It
```
Comment (webknjaz) [inline proxy/core/acceptor/threadless.py:57]: @abhinavsingh the leading tilde means that only the last chunk `Work` will be visible in HTML.

File: docs/conf.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+# There are two options for replacing |today|: either, you set today to some
+# non-false value, then it is used:
+# today = ''
+# Else, today_fmt is used as the format for a strftime call.
+today_fmt = '%B %d, %Y'
+
+# The reST default role (used for this markup: `text`) to use for all
+# documents.
+# Ref: python-attrs/attrs#571
+default_role = 'any'
```
Comment (webknjaz) [inline docs/conf.py:71]: @abhinavsingh this instructs Sphinx to link any text in single backticks to any role that has an object with that name. Practically, it allows you to identify forgotten syntax (Markdown mixup in RST). In RST, it's best to set roles explicitly.

File: docs/conf.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+# The master toctree document.
+root_doc = master_doc = 'index'  # Sphinx 4+ / 3-  # noqa: WPS429
+
+
+# -- Extension configuration -------------------------------------------------
+
+# -- Options for intersphinx extension ---------------------------------------
+
+intersphinx_mapping = {
```
Comment (webknjaz) [inline docs/conf.py:161]: @abhinavsingh intersphinx allows you to reference objects defined on other Sphinx sites for as long as you list them here. Sphinx will download their `objects.inv` files with lists of everything exposed.
The built-in way of exploring those objects is:
```console
$ python -m sphinx.ext.intersphinx https://docs.python.org/3/objects.inv
```

But it's not very user-friendly, so I have a small project displaying this information for several sites I sometimes interact with. It's here: https://webknjaz.github.io/intersphinx-untangled/.

When linking objects from a foreign site, you can type them in as per usual, with their correct role and name. But since objects with the same name may exist on this and other sites, I prefer explicitly predending the references with `<site-alias>:`. The dict keys below are those aliases you can use to point at objects at a specific site.

File: docs/conf.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+]
+
+# -- Options for extlinks extension ------------------------------------------
+
+extlinks = {
+    'issue': (f'{github_repo_url}/issues/%s', '#'),  # noqa: WPS323
+    'pr': (f'{github_repo_url}/pull/%s', 'PR #'),  # noqa: WPS323
+    'commit': (f'{github_repo_url}/commit/%s', ''),  # noqa: WPS323
+    'gh': (f'{github_url}/%s', 'GitHub: '),  # noqa: WPS323
+    'user': (f'{github_sponsors_url}/%s', '@'),  # noqa: WPS323
```
Comment (webknjaz) [inline docs/conf.py:206]: @abhinavsingh these are simple roles for making external references. For example, you can use `user` (in RST) in as follows:
```rst
:user:`webknjaz`
```
And that will link to my sponsors page (for users w/o sponsors enabled, GH will automatically redirect to their profiles).
This is useful to credit contributions to people in the changelog, for example.

File: docs/conf.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+# The reST default role (used for this markup: `text`) to use for all
+# documents.
+# Ref: python-attrs/attrs#571
+default_role = 'any'
+
+nitpicky = True
+_any_role = 'any'
+_py_class_role = 'py:class'
+nitpick_ignore = [
+    (_any_role, '<proxy.HttpProxyBasePlugin>'),
```
Comment (webknjaz) [inline docs/conf.py:245]: @abhinavsingh these are weird things mostly in single backticks found across the codebase. They all should be fixed one way or another, this is why they are ignored for now. But the action items are the same as for those linting violations for flake8.

File: docs/glossary.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+(_proxy_py_glossary)=
+# {{ project }} Glossary
```
Comment (webknjaz) [inline docs/glossary.md:2]: @abhinavsingh I declared this variable substitution so you can use it as `{{ project }}` in Markdown documents and as `|project|` in RST.

File: docs/glossary.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+```{glossary}
+DoH
+DNS over HTTP
+DNS-over-HTTP
+
+  [DNS over HTTP]
+```
+
+```{glossary}
+DNS
```
Comment (webknjaz) [inline docs/glossary.md:19]: @abhinavsingh these are declared mostly so that they could be used via a `term` directive instead of adding them to spelling exclusions. It's also useful for interlinking the same concept from various places. The lists may contain a few aliases as demonstrated for DoH. RST usage example:
```rst
:term:`DNS`
:term:`a custom DNS mention anchor <DNS>`
```

File: docs/index.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+```{include} ../README.md
+:start-after: (DO-NOT-REMOVE-docs-badges-END)
+```
+
+```{toctree}
+:hidden:
+
+Glossary <glossary>
```
Comment (webknjaz) [inline docs/index.md:31]: This syntax allows choosing a custom name for the menu item. In most cases, you'd just type in the document name and its title would be used here.

File: docs/index.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+```{include} ../README.md
+:end-before: (DO-NOT-REMOVE-docs-badges-START)
+```
+
+```{include} ../README.md
+:start-after: (DO-NOT-REMOVE-docs-badges-END)
+```
+
+```{toctree}
```
Comment (webknjaz) [inline docs/index.md:28]: @abhinavsingh you can have as many ToC trees as you want. Each may have a caption if needed. `:hidden:` is needed to exclude from the document render because this is also included in the sidebar and is unnecessary in the doc bottom.

File: docs/conf.py [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+github_url = 'https://github.com'
+github_repo_org = 'abhinavsingh'
+github_repo_name = 'proxy.py'
+github_repo_slug = f'{github_repo_org}/{github_repo_name}'
+github_repo_url = f'{github_url}/{github_repo_slug}'
+github_sponsors_url = f'{github_url}/sponsors'
+
+project = github_repo_name.title()
+author = f'{project} project contributors'
+copyright = author  # pylint: disable=redefined-builtin
```
Comment (webknjaz) [inline docs/conf.py:42]: I chose this style without a date or the names because of the recommendations from the Linux Foundation.
TL;DR it's easier to maintain and doesn't have direct legal implications anyway.

https://linuxfoundation.org/blog/copyright-notices-in-open-source-software-projects/

File: docs/glossary.md (hunk 11 above)
Comment (abhinavsingh) [reply docs/glossary.md:19]: All of this is awesomeness.  I did notice the glossary section :)

File: docs/spelling_wordlist.txt [hunk 15]
```diff
@@ -0,0 +1,7 @@
+IPv
```
Comment (webknjaz) [inline docs/spelling_wordlist.txt:1]: @abhinavsingh this could probably be moved to the glossary and linked as a term once it's out of the README.

File: tox.ini [hunk 16]
```diff
@@ -20,6 +20,139 @@ setenv =
   PEP517_OUT_DIR = {env:PEP517_OUT_DIR:{toxinidir}{/}dist}
 
 
+[testenv:build-docs]
+allowlist_externals =
+  git
+basepython = python3
+commands_pre =
+  # Paramiko:
+  {envpython} -m pip install -r{toxinidir}/requirements-tunnel.txt
```
Comment (webknjaz) [inline tox.ini:29]: This is not in the deps because the docs requirements use a constraints file with hashes and when pip sees at least one hash, it requires all of the packages to be listed with hashes. The one on this line doesn't have hashes and will fail if listed in the same command. It's a hack but it is what it is. If the env deps infra is properly converted to pip-tools managed constraints files, it would be possible to move this to a better place.

File: tox.ini [hunk 17]
```diff
@@ -20,6 +20,139 @@ setenv =
...
+  # -c{toxinidir}/docs/requirements.txt
+description = Build The Docs
+isolated_build = true
+passenv =
+  SSH_AUTH_SOCK
+skip_install = false
+usedevelop = false
+
+
+[testenv:doctest-docs]
```
Comment (webknjaz) [inline tox.ini:73]: @abhinavsingh with this, you can start adding doctests as part of the docs snippets per https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html. It's important to know that those snippets aren't out-of-date and still work.

**Comment by webknjaz** [comment]
@abhinavsingh you can now use the short link https://proxypy.rtfd.io on the repository front page, I think.

Also, I have a comment about the pictures. The image in the README is not transparent and is not very well combined with the Dark Mode. It is visible both on the docs site and on GitHub (if you switch to dark mode).

I recommend remaking that picture to:
1) Use a transparent background
2) Play well with both white page background and black/dark-gray (this may need experimentation to come up with good colors suitable for both cases simultaneously)
3) Make that image SVG, not PNG

As a side note, think about creating an SVG favicon too — try to avoid committing binary blobs like pictures into Git repos, it's harmful to them. But since SVG is just text/XML, it's fine + it's great for viewing with the 200% zoom that some people have in their browsers.
